overlord/assertstate/assertstate.go

```diff
@@ -889,10 +891,82 @@ func TryEnforceValidationSets(st *state.State, validationSets []string, userID i
 	return addCurrentTrackingToValidationSetsHistory(st)
 }
 
-// EnforceValidationSet tries to fetch the given validation set and enforce it.
+// EnforceValidationSets enforces the supplied validation sets. It takes a map
+// of validation set keys to validation sets, pinned sequence numbers (if any),
+// installed snaps and ignored snaps. It fetches any pre-requisites necessary.
+func EnforceValidationSets(st *state.State, valsets map[string]*asserts.ValidationSet, pinnedSeqs map[string]int, snaps []*snapasserts.InstalledSnap, ignoreValidation map[string]bool, userID int) error {
+	deviceCtx, err := snapstate.DevicePastSeeding(st, nil)
+	if err != nil {
+		return err
+	}
+
+	db := cachedDB(st)
+	batch := asserts.NewBatch(handleUnsupported(db))
+
+	vss := make([]*asserts.ValidationSet, 0, len(valsets))
+	valsetTracking := make([]*ValidationSetTracking, 0, len(valsets))
+
+	for vsKey, vs := range valsets {
+		pinnedSeq := pinnedSeqs[vsKey]
+		if pinnedSeq != 0 && pinnedSeq != vs.Sequence() {
+			// shouldn't be possible save for programmer error since, if we have a pinned
+			// sequence here, it should've been used when fetching the assertion
+			return fmt.Errorf("internal error: trying to enforce validation set %q with sequence point %d different than pinned %d", vsKey, vs.Sequence(), pinnedSeq)
+		}
+
+		tr := &ValidationSetTracking{
+			AccountID: vs.AccountID(),
+			Name:      vs.Name(),
+			Mode:      Enforce,
+			Current:   vs.Sequence(),
```

Comment:
this is not the problem of this PR and not a blocker for it but we need to look again how we define Current, we have places that set it to the enforced sequence point (pinned or not) and others that set it to the latest we have on the device, which is also the enforced one if not pinned but can be something else if pinned.